### PR TITLE
Gave SaveGroupingDefinition more input options

### PR DESCRIPTION
**Pull request by dlcaballero16**
## Description of work

This is to fix the issue of SaveGroupDefinition crashing when supplying an XML file.

## Explanation of work

Inputs were added to SaveGroupingDefinition that can be passed onto LoadGroupingDefiniton. The failure was happening because LoadGroupDefinition did not have InstrumentName specified.

### Dev testing

Make sure unit tests pass.

### CIS testing

Run the following script in the mantid workbench to verify. Running this before would result in a crash, but it will now finish executing. You should probably update localDir to be a local directory to you.

``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import pathlib
import os.path

from snapred.backend.recipe.algorithm.LoadGroupingDefinition import LoadGroupingDefinition as LoadingAlgo
from snapred.backend.recipe.algorithm.SaveGroupingDefinition import SaveGroupingDefinition as SavingAlgo
from snapred.meta.Config import Config, Resource

localDir = '/SNS/users/8l2/Documents/'
PGDDir = '/SNS/SNAP/shared/Calibration/Powder/PixelGroupingDefinitions/'
groupingFile = PGDDir + "SNAPFocGroup_Column.xml"

Config._config["cis_mode"] = False # <<-- change to True and rerun

output_file_name = pathlib.Path(groupingFile).stem + ".hdf"
outputFilePath = os.path.join(localDir, output_file_name)
# save the input grouping file in the calibration format
savingAlgo = SavingAlgo()
savingAlgo.initialize()
savingAlgo.setProperty("GroupingFilename", groupingFile)
savingAlgo.setProperty("OutputFilename", outputFilePath)
savingAlgo.setProperty("InstrumentName", "SNAP")
assert savingAlgo.execute()
```

## Link to EWM item

[EWM#2638](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=2638)

### CODE REVIEW

File: src/snapred/backend/recipe/algorithm/LoadGroupingDefinition.py [hunk 1]
```diff
@@ -111,6 +111,7 @@ def PyExec(self) -> None:
                 instrument_donor = self.mantidSnapper.LoadEmptyInstrument(
                     "Loading instrument definition file...",
                     Filename=self.getProperty("InstrumentFilename").value,
+                    InstrumentName=self.getProperty("InstrumentName").value,
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/LoadGroupingDefinition.py:114]: Here you can actually use all of `InstrumentName`, `InstrumentDonor`, and `Filename` and `LoadEmptyInstrument` should handle it

File: src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py [hunk 2]
```diff
@@ -62,6 +79,21 @@ def validateInput(self) -> None:
         if file_extension not in self.supported_calib_file_extensions:
             raise Exception(f"OutputFilename has an unsupported file name extension {file_extension}")
 
+        instrumentName = self.getProperty("InstrumentName").value
+        instrumentFileName = self.getProperty("InstrumentFileName").value
+        instrumentDonor = self.getProperty("InstrumentDonor").value
+        instrumentPropertiesCount = sum(not prop for prop in [instrumentName, instrumentFileName])
+        if instrumentDonor:
+            instrumentPropertiesCount += 1
```
[the remark this answers is not in the document]
Comment (dlcaballero16) [reply src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py:87]: I was having issues with putting instrumentDonor in the list. When I printed out the value of instrumentDonor, it was "None". So I think it was actually being set to the string "None" and not actually a object type of None.

File: src/snapred/backend/recipe/algorithm/LoadGroupingDefinition.py (hunk 1 above)
Comment (dlcaballero16) [reply src/snapred/backend/recipe/algorithm/LoadGroupingDefinition.py:114]: Are you sure I need InstrumentDonor? That's what this section is setting, and it only gets here if InstrumentDonor isn't set. I will add InstrumentFilename though

Comment (rboston628) [reply src/snapred/backend/recipe/algorithm/LoadGroupingDefinition.py:114]: You're right, I'm wrong.

File: src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rboston628) [reply src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py:87]: I would imagine it'd be set to `""`, but try `instrumentDonor = self.getProperty("InstrumentDonor")` without the `.value`

[the remark this answers is not in the document]
Comment (dlcaballero16) [reply src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py:87]: Looks like without `.value` it returns a reference to the instrumentDonor object

[the remark this answers is not in the document]
Comment (rboston628) [reply src/snapred/backend/recipe/algorithm/SaveGroupingDefinition.py]: Yes, it should be a reference

**Review by rboston628 — APPROVED** [review]

**Comment by rboston628** [comment]
I tested with the following script, and saw the save algo can run in all three cases:
``` python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import pathlib
import os.path

from snapred.backend.recipe.algorithm.LoadGroupingDefinition import LoadGroupingDefinition as LoadingAlgo
from snapred.backend.recipe.algorithm.SaveGroupingDefinition import SaveGroupingDefinition as SavingAlgo
from snapred.meta.Config import Config, Resource

PGDDir = '/SNS/SNAP/shared/Calibration_old/PixelGroupingDefinitions/'
groupingFile = PGDDir + "SNAPFocGrp_Column.lite.xml"
outputFilePath = "~/Documents/testoutput.hdf" 

## TEST RUNNING WITH INSTRUMENT NAME
savingAlgo = SavingAlgo()
savingAlgo.initialize()
savingAlgo.setProperty("GroupingFilename", groupingFile)
savingAlgo.setProperty("OutputFilename", outputFilePath)
savingAlgo.setProperty("InstrumentName", "SNAP")
assert savingAlgo.execute()

## TEST RUNNINNG WITH INSTRUMENT FILE
savingAlgo = SavingAlgo()
savingAlgo.initialize()
savingAlgo.setProperty("GroupingFilename", groupingFile)
savingAlgo.setProperty("OutputFilename", outputFilePath)
savingAlgo.setProperty("InstrumentFilename", "/SNS/SNAP/shared/NeXusFiles/SNAP_Definition.xml")
assert savingAlgo.execute()

## TEST RUNNING WITH INSTRUMENT DONOR
LoadEmptyInstrument(
    OutputWorkspace="_sampleWS",
    InstrumentName="SNAP",
)
savingAlgo = SavingAlgo()
savingAlgo.initialize()
savingAlgo.setProperty("GroupingFilename", groupingFile)
savingAlgo.setProperty("OutputFilename", outputFilePath)
savingAlgo.setProperty("InstrumentDonor", "_sampleWS")
assert savingAlgo.execute()
```

**Review by rboston628 — APPROVED** [review]

**Pull request merged**